WeakAuras/TimeMachine.lua

```diff
@@ -0,0 +1,299 @@
+function TimeMachine.actions.insert(data, path, payload)
+  local tbl, key = resolveKey(data, path)
+  if payload.index == nil then
+    table.insert(tbl[key], payload.value)
+  else
+    table.insert(tbl[key], payload.index, payload.value)
+  end
+end
+
+---@type Inverter<{index: number, value: any}, number>
+function TimeMachine.inverters.insert(data, path, payload)
+  return 'remove', path, payload.index
+end
+
+---@type Action<number | nil>
+function TimeMachine.actions.remove(data, path, payload)
+  local tbl, key = resolveKey(data, path)
+  if payload == nil then
+    table.remove(tbl[key])
+  else
+    table.remove(tbl[key], payload)
+  end
+end
+
+---@type Inverter<number, {index: number?, value: any}>
+function TimeMachine.inverters.remove(data, path, payload)
+  local tbl, key = resolveKey(data, path)
+  return 'insert', path, {index = payload, value = copy(tbl[key], payload or #tbl[key])}
+end
+
+---@type Action<{[1]: number, [2]: number}>
+function TimeMachine.actions.swap(data, path, payload)
+  local tbl, key = resolveKey(data, path)
+  tbl[key][payload[1]], tbl[key][payload[2]] = tbl[key][payload[2]], tbl[key][payload[1]]
+end
+
+---@type Inverter<{[1]: number, [2]: number}, {[1]: number, [2]: number}>
+function TimeMachine.inverters.swap(data, path, payload)
+  return 'swap', path, {payload[2], payload[1]}
+end
+
+---@type Action<{[1]: number, [2]: number}>
+function TimeMachine.actions.move(data, path, payload)
+  local tbl, key = resolveKey(data, path)
+  local value = table.remove(tbl, payload[1])
+  table.insert(tbl[key], payload[2], value)
+end
+
+---@type Inverter<{[1]: number, [2]: number}, {[1]: number, [2]: number}>
+function TimeMachine.inverters.move(data, path, payload)
+  return 'move', path, {payload[2], payload[1]}
+end
+
+---@param path keyPath
+local function keyPathToString(path)
+  if type(path) == 'table' then
+    return table.concat(path, '.')
+  else
+    return path
+  end
+end
+
+function TimeMachine:StartTransaction()
+  self:Reject()
+  self.transaction = true
+end
+
+---@param record actionRecord
+function TimeMachine:Append(record)
+  local action = self.actions[record.actionType]
+  Private.DebugPrint("Forward action", record.actionType, "for", record.uid, "at", keyPathToString(record.path), "with", record.payload)
+  if not action then
+    error("No action for actionType: " .. record.actionType)
+  end
+  local inverter = self.inverters[record.actionType]
+  if not inverter then
+    error("No inverter for action: " .. record.actionType)
+  end
+  local actionType, path, payload = inverter(Private.GetDataByUID(record.uid), record.path, record.payload)
+  local inverseRecord = {
+    uid = record.uid,
+    actionType = actionType,
+    path = path,
+    payload = payload
+  }
+  Private.DebugPrint("Backward action", actionType, "for", record.uid, "at", keyPathToString(path), "with", payload)
+  table.insert(self.next.forward, record)
+  table.insert(self.next.backward, inverseRecord)
```

Comment:
I think it might be good if the backward steps are taken in inverse order. That only changes anything if two steps depend on each other or modify the same value.